### ISSUE TITLE
[FR] VMC support

### ISSUE THREAD

**Issue by vivi90** [emilianavt/OpenSeeFace#38]
How about to add VMC Assistant support to OpenSeeFace?

I have read the [docu of the protocol](https://protocol.vmc.info/english).

Currently iam writing an python `pip` package for vmc.
But i have some questions about bone transform:
```
/VMC/Ext/Bone/Pos <boneName> <p.x> <p.y> <p.z> <q.x> <q.y> <q.z> <q.w>
```
After running an OSC sniffer with VSeeFace in "transmitter mode" and covered webcam i get for each bone the T-Pose values of my model.
That's clear.
Now the question: How do you apply the tracking data to these T-Pose to get the correct position and quaternion?

__In short:__
So if i want to add VMC support to OpenSeeFace, then how to apply the tracking data from the webcam to the initial T-Pose of the model bones?

**Comment by emilianavt** [comment]
Adding VMC protocol support to OpenSeeFace's face tracker itself would be difficult, because it requires the full skeleton from a 3D model, which in turn would require IK to animate from the head target, which is the only information the tracker itself actually provides.

If you are using the Unity components, you can just use this [BonesSend.cs](https://github.com/emilianavt/VSeeFacePreview/blob/master/Assets/VMC/BonesSend.cs) in combination with [OpenSeeFaceSample](https://github.com/emilianavt/OpenSeeFaceSample/). But since you mentioned writing a `pip` package, that's probably not an option.

**Comment by emilianavt** [comment]
You just set the rotations and positions you receive to every bone's rotation value. For sending, you take each bone's rotation and position value and send it.

To apply the face tracking's head position and pose to the model however, you need IK.

**Comment by vivi90** [comment]
Here ist the corresponding feature request for [ROMP](https://github.com/Arthur151/ROMP):
https://github.com/Arthur151/ROMP/issues/193

**Comment by emilianavt** [comment]
> For me it seems, that the bone values are based on the model.
> So can i really just apply the tracking results for each bone?

It depends on the model. For VRM models, all bone rotations are normalized to 0 while in a T pose. Bone positions will differ of course, but usually only applying rotations works fine. If you are working with unnormalized models, things will become much trickier and you'll have to figure out a way of rotating things.

If you are sending the head pose as a tracker to VMC itself, it should handle the IK for you and send you bone rotations.

**Comment by vivi90** [comment]
> > For me it seems, that the bone values are based on the model.
> > So can i really just apply the tracking results for each bone?
> 
> [..] For VRM models, all bone rotations are normalized to 0 while in a T pose [..] applying rotations works fine [..]

Okay, i will test it 🙂

> If you are sending the head pose as a tracker to VMC itself, it should handle the IK for you and send you bone rotations.

Do mean the official VMC software?
It's a little bit confusing, that both the protocol and the software is called `VMC` ^^

**Comment by emilianavt** [comment]
> Do mean the official VMC software?

The protocol is called "VMC protocol", so I assumed the box labelled just "VMC" was referring to the software.

**Comment by vivi90** [comment]
> If you are sending the head pose as a tracker to VMC itself, it should handle the IK for you and send you bone rotations.

Just to clarify the things:
According to the https://protocol.vmc.info/english#glossary it seems we don't need IK for VMC protocol support at all in OpenSeeFace, if we only implement it as an 'VMC Assistent'.
Also it might anyway be used together with 'VMC Performer' (IK supporting) applications like VSeeFace or VMC software.

So i will create an PR, if my pip package is usable.

**Comment by vivi90** [comment]
If somebody else reads this thread later and is looking for human IK in python, maybe this helps: https://github.com/nghorbani/human_body_prior

**Comment by vivi90** [comment]
> [..] Bone positions will differ of course, but usually only applying rotations works fine [..]

Which minimum and maximum ranges are typical? 🙂 
Because `[-180.0, 180.0]` seems to huge and `[-1.0, 1.0]` seems small.

**Comment by emilianavt** [comment]
> According to the https://protocol.vmc.info/english#glossary it seems we don't need IK for VMC protocol support at all in OpenSeeFace, if we only implement it as an 'VMC Assistent'.

You can send the head target as a `/VMC/Ext/Tra`, in that case you don't need IK, but not all applications will be able to handle the received data/perform IK on VMC protocol tracker data. For example, I believe VSeeFace ignores received `/VMC/Ext/Tra` data. If you want to apply it to the bones and send `/VMC/Ext/Bone`, you do need IK.

> Which minimum and maximum ranges are typical?

They are quaternions, so there isn't really a minimum or maximum range, especially if un-normalized.
